File: services/api-gateway/src/handlers/booking/cancel.ts
import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { AppError, ErrorCode, BookingStatus, VALID_STATUS_TRANSITIONS } from '@towcommand/core';
import { calculateCancellationFee } from '@towcommand/core';
import { BookingRepository } from '@towcommand/db';
import { publishEvent, EVENT_CATALOG } from '@towcommand/events';
import { handleError, successResponse } from '../../middleware/error-handler';

const bookingRepo = new BookingRepository();

export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  try {
    const userId = event.requestContext.authorizer?.userId as string;
    const bookingId = event.pathParameters?.id as string;
    const body = JSON.parse(event.body ?? '{}');
    const reason = body.reason as string | undefined;

    const booking = await bookingRepo.getById(bookingId);

    if (!booking) {
      throw AppError.notFound('Booking');
    }

    if (booking.customerId !== userId) {
      throw AppError.forbidden('You can only cancel your own bookings');
    }

    const allowedTransitions = VALID_STATUS_TRANSITIONS[booking.status];
    if (!allowedTransitions.includes(BookingStatus.CANCELLED)) {
      throw AppError.badRequest(
        ErrorCode.CANNOT_CANCEL,
        `Cannot cancel booking in ${booking.status} status`,
      );
    }

    const cancellationFee = calculateCancellationFee(booking.status, 0);

    await bookingRepo.updateStatus(bookingId, BookingStatus.CANCELLED, {
      cancelledBy: userId,
      reason,
      cancellationFee,
    });

    await publishEvent(
      EVENT_CATALOG.booking.source,
      EVENT_CATALOG.booking.events.BookingCancelled,
      {
        bookingId,
        customerId: userId,
        providerId: booking.providerId,
        previousStatus: booking.status,
        reason,
        cancellationFee,
      },
      { userId, userType: 'customer' },
    );

    return successResponse({
      bookingId,
      status: BookingStatus.CANCELLED,
      cancellationFee,
    });
  } catch (error) {
    return handleError(error);
  }
}

File: internal/usecase/booking/cancel.go
package bookinguc

import (
	"context"

	"github.com/davidlramirez95/towcommand/internal/domain/booking"
	domainerrors "github.com/davidlramirez95/towcommand/internal/domain/errors"
	"github.com/davidlramirez95/towcommand/internal/domain/user"
)

const (
	eventBookingCancelled = "BookingCancelled"
)

// CancelBookingInput holds the data needed to cancel a booking.
type CancelBookingInput struct {
	BookingID string
	CallerID  string
	Reason    string
}

// CancelBookingOutput is the response for a cancel operation.
type CancelBookingOutput struct {
	BookingID       string `json:"bookingId"`
	Status          string `json:"status"`
	CancellationFee int64  `json:"cancellationFee"`
}

// CancelBookingRepo combines the interfaces needed by the cancel use case.
type CancelBookingRepo interface {
	BookingFinder
	BookingStatusUpdater
}

// CancelBookingUseCase orchestrates booking cancellation.
type CancelBookingUseCase struct {
	repo   CancelBookingRepo
	events EventPublisher
}

// NewCancelBookingUseCase constructs a CancelBookingUseCase with its dependencies.
func NewCancelBookingUseCase(repo CancelBookingRepo, events EventPublisher) *CancelBookingUseCase {
	return &CancelBookingUseCase{repo: repo, events: events}
}

// Execute cancels a booking, enforcing ownership and valid state transitions.
func (uc *CancelBookingUseCase) Execute(ctx context.Context, input CancelBookingInput) (*CancelBookingOutput, error) {
	b, err := uc.repo.FindByID(ctx, input.BookingID)
	if err != nil {
		return nil, err
	}
	if b == nil {
		return nil, domainerrors.NewNotFoundError("Booking", input.BookingID)
	}

	if b.CustomerID != input.CallerID {
		return nil, domainerrors.NewForbiddenError("You can only cancel your own bookings")
	}

	if !booking.CanTransition(b.Status, booking.BookingStatusCancelled) {
		return nil, domainerrors.NewBookingNotCancellableError(input.BookingID, string(b.Status))
	}

	cancellationFee := calculateCancellationFee(b.Status)

	metadata := map[string]any{
		"cancelledBy":     input.CallerID,
		"reason":          input.Reason,
		"cancellationFee": cancellationFee,
	}
	if err := uc.repo.UpdateStatus(ctx, input.BookingID, booking.BookingStatusCancelled, metadata); err != nil {
		return nil, err
	}

	_ = uc.events.Publish(ctx, eventSourceBooking, eventBookingCancelled, map[string]any{
		"bookingId":       input.BookingID,
		"customerId":      b.CustomerID,
		"providerId":      b.ProviderID,
		"previousStatus":  b.Status,
		"reason":          input.Reason,
		"cancellationFee": cancellationFee,
	}, &Actor{UserID: input.CallerID, UserType: string(user.UserTypeCustomer)})

	return &CancelBookingOutput{
		BookingID:       input.BookingID,
		Status:          string(booking.BookingStatusCancelled),
		CancellationFee: cancellationFee,
	}, nil
}

// calculateCancellationFee returns the fee in centavos based on current status.
func calculateCancellationFee(status booking.BookingStatus) int64 {
	switch status {
	case booking.BookingStatusPending:
		return 0
	case booking.BookingStatusMatched:
		return 10_000 // ₱100
	case booking.BookingStatusEnRoute:
		return 25_000 // ₱250
	default:
		return 0
	}
}
